File: bot/crypto.py
"""CryptoJS 호환 AES 복호화 모듈

seven-split의 crypto.ts와 동일한 방식으로 복호화
CryptoJS.AES.encrypt()는 OpenSSL 호환 형식 사용
"""
import base64
import hashlib
from typing import Optional
from Crypto.Cipher import AES
from Crypto.Util.Padding import unpad


def decrypt_aes(encrypted: str, passphrase: str) -> str:
    """CryptoJS AES 암호화 문자열 복호화

    CryptoJS는 OpenSSL 호환 형식 사용:
    - Base64(Salted__ + 8바이트 salt + 암호문)
    - 키/IV는 passphrase + salt로 파생
    """
    if not encrypted or not passphrase:
        return ""

    try:
        # Base64 디코딩
        data = base64.b64decode(encrypted)

        # "Salted__" 헤더 확인 (8바이트)
        if data[:8] != b"Salted__":
            # 레거시 Caesar 암호 시도
            return _decrypt_legacy(encrypted)

        # Salt 추출 (8바이트)
        salt = data[8:16]
        ciphertext = data[16:]

        # 키와 IV 파생 (OpenSSL EVP_BytesToKey 방식)
        key, iv = _evp_bytes_to_key(passphrase.encode(), salt, 32, 16)

        # AES-256-CBC 복호화
        cipher = AES.new(key, AES.MODE_CBC, iv)
        decrypted = unpad(cipher.decrypt(ciphertext), AES.block_size)

        return decrypted.decode('utf-8')

    except Exception as e:
        print(f"[Crypto] AES 복호화 실패: {e}")
        # 레거시 시도
        try:
            return _decrypt_legacy(encrypted)
        except:
            return ""


def _evp_bytes_to_key(password: bytes, salt: bytes, key_len: int, iv_len: int) -> tuple[bytes, bytes]:
    """OpenSSL EVP_BytesToKey 구현 (MD5 기반)"""
    d = b""
    d_i = b""

    while len(d) < key_len + iv_len:
        d_i = hashlib.md5(d_i + password + salt).digest()
        d += d_i

    return d[:key_len], d[key_len:key_len + iv_len]


def _decrypt_legacy(encoded: str) -> str:
    """레거시 Caesar 암호 복호화 (마이그레이션용)"""
    SHIFT = 7
    try:
        shifted = base64.b64decode(encoded).decode('utf-8')
        return ''.join(chr(ord(c) - SHIFT) for c in shifted)
    except:
        return ""

File: src/utils/crypto.ts
// AES 암호화 (crypto-js 사용)
import CryptoJS from 'crypto-js';

const getEncryptionKey = (): string => {
  const key = import.meta.env.VITE_ENCRYPTION_KEY;
  if (!key) {
    throw new Error('암호화 키 환경변수가 설정되지 않았습니다.');
  }
  return key;
};

export function encrypt(text: string): string {
  if (!text) return '';
  try {
    const key = getEncryptionKey();
    return CryptoJS.AES.encrypt(text, key).toString();
  } catch (e) {
    console.error('암호화 실패:', e);
    return '';
  }
}

export function decrypt(encoded: string): string {
  if (!encoded) return '';

  // 먼저 기존 Caesar 암호로 시도 (마이그레이션용)
  // Caesar 암호는 Base64로 시작하므로 U2Fsd로 시작하지 않으면 Caesar일 가능성 높음
  if (!encoded.startsWith('U2Fsd')) {
    try {
      const legacy = decryptLegacy(encoded);
      if (legacy) return legacy;
    } catch {
      // Caesar 실패, AES 시도
    }
  }

  // AES 복호화 시도
  try {
    const key = getEncryptionKey();
    const bytes = CryptoJS.AES.decrypt(encoded, key);
    const decrypted = bytes.toString(CryptoJS.enc.Utf8);
    if (decrypted) return decrypted;
  } catch {
    // AES 실패
  }

  // 둘 다 실패하면 빈 문자열
  console.warn('복호화 실패:', encoded.substring(0, 20) + '...');
  return '';
}

// 기존 Caesar 암호 복호화 (마이그레이션용)
function decryptLegacy(encoded: string): string {
  const SHIFT = 7;
  const shifted = atob(encoded);
  return shifted
    .split('')
    .map((char) => String.fromCharCode(char.charCodeAt(0) - SHIFT))
    .join('');
}
